src/main/java/nl/knaw/dans/pf/language/ddm/handlers/EasSpatialHandler.java

```diff
@@ -67,22 +108,77 @@ private void checkSRS(final Attributes attributes) {
+                interiorDescription = getCharsSinceStart().trim();
+            else
+                description = getCharsSinceStart().trim();
+        } else if ("pos".equals(localName))
             pos = createPoint();
         else if ("lowerCorner".equals(localName))
             lower = createPoint();
         else if ("upperCorner".equals(localName))
             upper = createPoint();
-        else if ("Point".equals(localName)) {
-            if (pos != null)
-                getTarget().getEmdCoverage().getEasSpatial().add(new Spatial(description, pos));
-        } else if ("Envelope".equals(localName)) {
-            if (lower != null && upper != null)
-                getTarget().getEmdCoverage().getEasSpatial().add(new Spatial(description, createBox()));
+        else if ("posList".equals(localName)) {
+            if (state == E_POSLIST) {
+                exteriorPoints = createPolygonPoints();
+                state = state.getNextState();
+            } else if (state == I_POSLIST) {
+                interiorPoints = createPolygonPoints();
+                state = state.getNextState();
+            }
+        } else if ("exterior".equals(localName) && state == END_EXTERIOR) {
+            exteriorPart = new PolygonPart(exteriorDescription, exteriorPoints);
+            state = state.getNextState();
+        } else if ("interior".equals(localName) && state == END_INTERIOR) {
+            interiorParts.add(new PolygonPart(interiorDescription, interiorPoints));
+            state = state.getNextState();
+        } else if ("Polygon".equals(localName) && state == END_POLYGON) {
+            Spatial.Polygon polygon = createPolygon();
+            getTarget().getEmdCoverage().getEasSpatial().add(new Spatial(polygonDescription, polygon));
+            state = state.getNextState();
+        } else if ("Point".equals(localName) && pos != null)
+            getTarget().getEmdCoverage().getEasSpatial().add(new Spatial(description, pos));
+        else if ("Envelope".equals(localName) && lower != null && upper != null)
+            getTarget().getEmdCoverage().getEasSpatial().add(new Spatial(description, createBox()));
+        // other types than point/box/polygon not supported by EMD: don't warn
+    }
+
+    private List<PolygonPoint> createPolygonPoints() throws SAXException {
+        String[] coordinates = getCharsSinceStart().trim().split(" ");
+        int length = coordinates.length;
+        if (length < 8) {
+            error("expected at least 4 coordinate pairs to construct at least a triangle");
+            return null;
+        } else if (length % 2 == 1) {
+            error("expected an even number of coordinates since they're taken in pairs of two");
+            return null;
+        } else if (!coordinates[0].equals(coordinates[length - 2]) && !coordinates[1].equals(coordinates[length - 1])) {
+            error("first pair of coordinates should equal the last pair of coordinates");
```

Comment:
Don't know how far validation should go. With five points you can easily define a bow tie. If that should not be  allowed check if lines cross on another, of course neighbors can't cross but you might not want to have them exactly on top. Interior polygons with lines crossing lines of exterior polygon would also be an error.

Saw you playing with bounding boxes, but a bounding box of the interior polygon inside the bounding box of the exterior polygon is not enough proof the interior is indeed inside the exterior.

Reply:
I didn't find any info on crossing lines. AFAIK GML doesn't state anything about it. Also about the 'interior being inside the exterior' that isn't tested for by GML. AFAIK they also don't state anything about that. Besides, I can easily think of [a case](https://www.google.nl/maps/place/Baarle-Hertog,+Belgi%C3%AB/@51.4242508,4.8546348,13z/data=!3m1!4b1!4m5!3m4!1s0x47c6a54331164f29:0x40099ab2f4d4d20!8m2!3d51.441832!4d4.931736?hl=nl) where multiple 'exteriors' have to be defined; even though we are only allowed (by GML) to have one exterior. I think that's where you can use these interiors for as well.

Also, while [playing with polygons](https://github.com/rvanheest/Experiments/blob/c08a5683c88562db57b2eae82c7457955ddab0c2/src/main/web/google_maps/polygonWithHoles.html#L35-L58) myself, I found that it is extremely important to have the order of the points (going clockwise vs counter-clockwise) correct. If you have a clockwise exterior and a clockwise interior, the interior will not be drawn as a 'hole'. For that you need a counter-clockwise interior... I am not sure whether this is a Google Maps specific thing or something more general, but GML doesn't state anything about that AFAIK, so I don't want to spend much time on these extra validations. It's up to the depositor to deliver its data correctly. Likewise, we don't do a spelling check on the description...